tload.1
.\"             -*-Nroff-*-
.\"  This page Copyright (C) 1993 Matt Welsh, mdw@tc.cornell.edu.
.\"  Freely distributable under the terms of the GPL
.TH TLOAD "1" "June 2011" "procps-ng" "User Commands"
.SH NAME
tload \- graphic representation of system load average
.SH SYNOPSIS
.B tload
[\fIoptions\fR] [\fItty\fR]
.SH DESCRIPTION
.B tload
prints a graph of the current system load average to the specified
.I tty
(or the tty of the tload process if none is specified).
.SH OPTIONS
.TP
\fB\-s\fR, \fB\-\-scale\fR \fInumber\fR
The scale option allows a vertical scale to be specified for the display (in
characters between graph ticks); thus, a smaller value represents a larger
scale, and vice versa.
.TP
\fB\-d\fR, \fB\-\-delay\fR \fIseconds\fR
The delay sets the delay between graph updates in
.IR seconds .
.TP
\fB\-h\fR, \fB\-\-help\fR
Display this help text.
.TP
\fB\-V\fR, \fB\-\-version\fR
Display version information and exit.
.PP
.SH FILES
.I /proc/loadavg
load average information
.SH "SEE ALSO"
.BR ps (1),
.BR top (1),
.BR uptime (1),
.BR w (1)
.SH BUGS
The
.BI "\-d" " delay"
option sets the time argument for an
.BR alarm (2);
if \-d 0 is specified, the alarm is set to 0, which will never send the
.B SIGALRM
and update the display.
.SH AUTHORS
Branko Lankester,
.UR david@\:ods.\:com
David Engel
.UE , and
.UR johnsonm@\:redhat.\:com
Michael K. Johnson
.UE .
.SH "REPORTING BUGS"
Please send bug reports to
.UR procps@freelists.org
.UE


NEWS
procps-ng-3.3.3
---------------
  * watch -g command repeats until something changes
  * Changed labels of kB etc to KiB Debian #662786
  * top supports hotplugged memory and cpus Debian #351934, #459287
  * top %cpu not distorted when switch to/from threads Debian #441166
  * top can read old toprc configuration files Debian #651213, #651863
  * top supports input paste once again Debian #663334
  * top -p avoids SEGV when pid doesn't exist Debian 668335

  * Documentation catch up (fixed prior to 3.3.3)
    . top repeating "%" wastes space Debian #322984
    . top -U user name parse error Debian #623200
    . top memory leaks Debian #627257
    . top impossible task swap statistics Debian #628462
    . top forest view segfault Debian #650864

procps-ng-3.3.1 --> procps-ng-3.3.2
--------------------------------

  * Redefined library to use version-info
  * NLS added! Programs can use po files to localise the output to suite
    any supported languages.
  * Imported a bunch of distribution patches:
      watch: support unicode
      watch: add precision wait time option -p
      watch: interpret ANSI color code sequences
      watch: add -exec and -beep flags and has better quoting
      w: use environment to set user and from/host column widths
      w: use COLUMNS environment if TIOCGWINSZ fails
      w: bassman emulation with -o option
      vmstat: do not scale si/so just like bi/bo
      libprocps-ng: sysinfo.c: truncate the vmstat figure to 32 bits
      tload: remote unneeded optarg and optind variables
      sysctl: fix up some option processing
      skill: kill prints perror
      skill: do not treat skill null parameter as 0
      skill: fix too greedy option parser
      libprocps-ng: readproc.c: some type conversion help
      ps: rename SZ to SIZE
      ps: add sorting to %mem for ps
      pmap: provide information for -x option
      pgrep: distinguish between invalid commandline parameters and '-?'
      pgrep: fix compiler warning saved_start_time might be used uninitialized
      pgrep: add -c option for counting number of matched proceesses
      pwdx & libprocps-ng: Hurd does not have MAX_PATH defined
      ps: --sort does not work with time argument
      skill: add CR to warning line
      contrib: minimal ps: define mips PAGE_SIZE
      libproc-ng: prettyfy proc mount messages
      ps: add build option to disable ps option warning
      libproc-ng: support building without WCHAR support
      sysctl: remove index() for buildroot
  * top now has a search capability
  * User import is validated better to stop negative values for intervals etc

procps-ng-3.3.0 --> procps-ng-3.3.1
--------------------------------

  * Added DejaGNU regression testing
  * Fixed pgrep -u not finding processes Debian#649591
  * Fixed pgrep crashing
  * vmstat -p <part> finds partitions. Was Debian patch vmstat_part_format
    fixes closed bugs RH#485243 and Debian#588677
  * watch 8-bit clean, Was Debian patch watch_8bitchar
  * slabtop prints plain ASCII in once mode

procps-3.2.8 --> procps-ng-3.3.0
------------------------------------------------------
Debian, Fedora and openSUSE fork of procps.
https://gitorious.org/procps

environment/build changes:
 . autotools integrated
 . procps-ng specific options
     --enable-oomem  (add out-of-memory flds)
     --disable-kill  (do not build kill pgm)
     --enable-w-from (from as default for w)

library changes:
 . added control groups support
 . added supplementary groups support
 . introduced proc_t memory management
 . enabled hot-plugged cpu support
 . introduced new api's
     readeither   (readproc.c)
     readproctab3 (readproc.c)
     escaped_copy (escape.c)

free changes:
 . rewritten/modernized
     improved help
     includes long options
     added new processing options
 . updated manual page

ps changes:
 . added new fields
     cgroups,etimes,supgid,supgrp,tgid
 . exploited new library api's

top changes:
 . redesigned fields management
 . added new fields
     cgroups,gid,nmin,nth,pgrp,ruid,
     sid,suid,supgids,supgrps,suser,
     tgid,tpgid
 . added horizontal/vertical scrolling
 . added flexible output override via -w
 . added COMMAND column forest view via V
 . added hot-plugged cpu/memory support
 . improved user filtering via u|U
 . improved signal handling
 . restored zero delay functionality
 . introduced true line input editing
 . exploited new library api's
 . improved performance up to 300%

Debian, Fedora and openSUSE fork of procps.
https://gitorious.org/procps
------------------------------------------------------

procps-3.2.8 --> procps-3.2.9 (pending)

top: fix an aliasing problem -- thanks David Owen

procps-3.2.7 --> procps-3.2.8

ps: allow "+" in sort specifications, as in man page     rh208217
ps: recognize SCHED_ISO and SCHED_IDLE
ps: document SCHED_BATCH and add a "see also" for stime
ps: man page less ambiguous
top: normal exit code should be 0     #341272 #354255 rh199174 suse493210
top: misc fixes
pgrep: usage error should exit with 2       #413383
vmstat: use EXIT_FAILURE -- thanks Yoshio Nakamura   #425492
sysctl: fix crash -- thanks Steinar Gunderson      #423704
watch: tolerate umlauts                          #207103
pmap: range limits with -A low,high
update /dev/tty* info to May 2009 devices.txt
don't read off end of string const                rh469495 rh498182

procps-3.2.6 --> procps-3.2.7

top: document H option -- thanks Tony Ernst
top: terabytes -- thanks Tony Ernst
top: CPU usage column width fixes -- thanks Tony Ernst
top: *roff change                                   #326517
ps: SCHED_BATCH is B
ps: fix s format (signals) output with thread display
watch: avoid integer overflow for the time delay
pwdx: buffer overflow fixed -- thanks Ulf Harnhammar
procps.spec needed a slash -- thanks Jesse Brandeburg
w: stale utmp entries snuck in via uninitialized var -- thanks Robert A Basch
pgrep/pkill: fix some realloc-related crashes       #353894

tload.c
/*
 * tload.c	- terminal version of xload
 *
 * Options:
 *	-s initial scaling exponent (default = 6)
 *	-d delay
 *
 * Copyright (c) 1992 Branko Lankester
 * /proc changes by David Engel (david@ods.com)
 * Made a little more efficient by Michael K. Johnson (johnsonm@sunsite.unc.edu)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include "proc/version.h"
#include "proc/sysinfo.h"
#include "c.h"
#include "fileutils.h"
#include "nls.h"
#include "strutils.h"
#include "xalloc.h"

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <setjmp.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <termios.h>
#include <unistd.h>

static char *screen;

static int nrows = 25;
static int ncols = 80;
static int scr_size;
static int fd = 1;
static unsigned int dly = 5;
static jmp_buf jb;

static void alrm(int signo __attribute__ ((__unused__)))
{
	signal(SIGALRM, alrm);
	alarm(dly);
}

static void setsize(int i)
{
	struct winsize win;

	signal(SIGWINCH, setsize);
	if (ioctl(fd, TIOCGWINSZ, &win) != -1) {
		if (win.ws_col > 0)
			ncols = win.ws_col;
		if (win.ws_row > 0)
			nrows = win.ws_row;
	}
	scr_size = nrows * ncols;
	if (screen == NULL)
		screen = (char *)xmalloc(scr_size);
	else
		screen = (char *)xrealloc(screen, scr_size);

	memset(screen, ' ', scr_size - 1);
	*(screen + scr_size - 2) = '\0';
	if (i)
		longjmp(jb, 0);
}

static void __attribute__ ((__noreturn__)) usage(FILE * out)
{
	fputs(USAGE_HEADER, out);
	fprintf(out,
	      _(" %s [options] [tty]\n"), program_invocation_short_name);
	fputs(USAGE_OPTIONS, out);
	fputs(_(" -d, --delay <secs>  update delay in seconds\n"
	        " -s, --scale <num>   vertical scale\n"), out);
	fputs(USAGE_SEPARATOR, out);
	fputs(USAGE_HELP, out);
	fputs(USAGE_VERSION, out);
	fprintf(out, USAGE_MAN_TAIL("tload(1)"));

	exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

int main(int argc, char **argv)
{
	int lines, row, col = 0;
	int i, opt;
	double av[3];
	static double max_scale = 0, scale_fact;
	long tmpdly;

	static const struct option longopts[] = {
		{"scale", required_argument, NULL, 's'},
		{"delay", required_argument, NULL, 'd'},
		{"help", no_argument, NULL, 'h'},
		{"version", no_argument, NULL, 'V'},
		{NULL, 0, NULL, 0}
	};
 
	program_invocation_name = program_invocation_short_name;
	setlocale (LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	atexit(close_stdout);

	while ((opt =
		getopt_long(argc, argv, "s:d:Vh", longopts, NULL)) != -1)
		switch (opt) {
		case 's':
			max_scale = strtod_or_err(optarg, _("failed to parse argument"));
			if (max_scale < 0)
			        xerrx(EXIT_FAILURE, _("scale cannot be negative"));
			break;
		case 'd':
			tmpdly = strtol_or_err(optarg, _("failed to parse argument"));
			if (tmpdly < 1)
				xerrx(EXIT_FAILURE, _("delay must be positive integer"));
			else if (UINT_MAX < tmpdly)
				xerrx(EXIT_FAILURE, _("too large delay value"));
			dly = tmpdly;
			break;
		case 'V':
			printf(PROCPS_NG_VERSION);
			return EXIT_SUCCESS;
			break;
		case 'h':
			usage(stdout);
		default:
			usage(stderr);
		}

	if (argc > optind)
		if ((fd = open(argv[optind], 1)) == -1)
			xerr(EXIT_FAILURE, _("can not open tty"));

	setsize(0);

	if (max_scale == 0)
		max_scale = nrows;

	scale_fact = max_scale;

	setjmp(jb);
	col = 0;
	alrm(0);

	while (1) {

		if (scale_fact < max_scale)
			scale_fact *= 2.0;	/* help it drift back up. */

		loadavg(&av[0], &av[1], &av[2]);

		while (1) {
			lines = av[0] * scale_fact;
			row = nrows - 1;

			while (0 <= --lines) {
				*(screen + row * ncols + col) = '*';
				if (--row < 0) {
					scale_fact /= 2.0;
					break;
				}
			}
			if (0 < row)
				break;
		}

		while (row >= 0)
			*(screen + row-- * ncols + col) = ' ';

		for (i = 1;; ++i) {
			char *p;
			row = nrows - (i * scale_fact);
			if (row < 0)
				break;
			if (*(p = screen + row * ncols + col) == ' ')
				*p = '-';
			else
				*p = '=';
		}

		if (++col == ncols) {
			--col;
			memmove(screen, screen + 1, scr_size - 1);

			for (row = nrows - 2; row >= 0; --row)
				*(screen + row * ncols + col) = ' ';
		}
		i = sprintf(screen, " %.2f, %.2f, %.2f", av[0], av[1], av[2]);
		if (i > 0)
			screen[i] = ' ';

		if (write(fd, "\033[H", 3) < 0)
			xerr(EXIT_FAILURE, _("writing to tty failed"));
		if (write(fd, screen, scr_size - 1) < 0)
			xerr(EXIT_FAILURE, _("writing to tty failed"));
		pause();
	}
}
